src/main/utility/synchronization/thread_parking.rs

```diff
@@ -0,0 +1,104 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+/// Used to unpark a thread, but which hasn't been assigned a specific thread yet.
+#[derive(Debug, Clone)]
+pub struct ThreadUnparkerUnassigned {
+    ready_flag: Arc<AtomicBool>,
+}
+
+/// Used to unpark a thread.
+#[derive(Debug, Clone)]
+pub struct ThreadUnparker {
+    thread: std::thread::Thread,
+    ready_flag: Arc<AtomicBool>,
+}
+
+/// Used to park a thread. The `ThreadParker` is derived from a `ThreadUnparker` or
+/// `ThreadUnparkerUnassigned`, and must only be used on the thread which the unparker was assigned
+/// to. If the `ThreadUnparker` was assigned to thread A, then `ThreadParker::park()` must only be
+/// called from thread A.
+#[derive(Debug, Clone)]
+pub struct ThreadParker {
+    ready_flag: Arc<AtomicBool>,
+}
+
+impl ThreadUnparkerUnassigned {
+    pub fn new() -> Self {
+        Self {
+            ready_flag: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    /// Assign this to a thread that will be unparked.
+    #[must_use]
+    pub fn assign(self, thread: std::thread::Thread) -> ThreadUnparker {
+        ThreadUnparker::new(self.ready_flag, thread)
+    }
+
+    /// Get a new [`ThreadParker`]. The `ThreadParker` must only be used from the thread which we
+    /// will later assign ourselves to using `assign()`. This is useful if you want to pass a
+    /// `ThreadParker` to a new thread before you have a handle to that thread.
+    pub fn parker(&self) -> ThreadParker {
+        ThreadParker::new(Arc::clone(&self.ready_flag))
+    }
+}
+
+impl ThreadUnparker {
+    fn new(ready_flag: Arc<AtomicBool>, thread: std::thread::Thread) -> Self {
+        Self { ready_flag, thread }
+    }
+
+    /// Unpark the assigned thread.
+    pub fn unpark(&self) {
+        self.ready_flag.store(true, Ordering::Release);
+        self.thread.unpark();
+    }
+
+    /// Get a new [`ThreadParker`] for the assigned thread.
+    pub fn parker(&self) -> ThreadParker {
+        ThreadParker::new(Arc::clone(&self.ready_flag))
+    }
+}
+
+impl ThreadParker {
+    fn new(ready_flag: Arc<AtomicBool>) -> Self {
+        Self { ready_flag }
+    }
+
+    /// Park the current thread until [`ThreadUnparker::unpark()`] is called. You must only call
+    /// `park()` from the thread which the corresponding `ThreadUnparker` is assigned, otherwise a
+    /// deadlock may occur.
+    pub fn park(&self) {
+        while self
+            .ready_flag
+            .compare_exchange(true, false, Ordering::SeqCst, Ordering::Relaxed)
```

Comment:
I'm not sure `SeqCst` makes sense here; I *think* it's only meaningful if other relevant operations are `SeqCst`, and that it might not be needed here (depending how the potential deadlock is resolved)

Reply:
I *think* this `SeqCst` can be `Relaxed`. Up to you, but maybe worth a comment about why you don't think it can be, or a `TODO` to further investigate if you want to just lean towards being more conservative for the moment.

Reply:
I agree that `SeqCst` isn't the right option here. What I want to prevent is any reordering with respect to the park() and unpark(), but I guess the memory orderings only synchronize what happens before the atomic operation, and there's nothing before the atomic operation that we need to synchronize. And here I think we have to hope/assume that the park() and unpark() operations have full memory barriers so that the prior atomic reads and writes will happen before the park() and unpark()?

Reply:
I think `Release` might make sense here. It's also free on x86-64, and what they use in the corresponding part of the example: https://doc.rust-lang.org/std/thread/fn.park.html#examples

I agree though that it feels like it's depending a little on the implementation details of park and unpark. It wouldn't hurt to throw in a compiler fence to ensure the atomic operation doesn't get moved to after the `park`. https://doc.rust-lang.org/std/sync/atomic/fn.compiler_fence.html

Reply:
Looking at the rust source, it has comments that say:

> So, since this parker is just a single atomic variable, we only need to look at the ordering guarantees we need to provide to the 'outside world'.
>
> The only memory ordering guarantee that parking and unparking provide, is that things which happened before unpark() are visible on the thread returning from park() afterwards. Otherwise, it was effectively unparked before unpark() was called while still consuming the 'token'.
>
> In other words, unpark() needs to synchronize with the part of park() that consumes the token and returns.

https://github.com/rust-lang/rust/blob/master/library/std/src/sys_common/thread_parker/futex.rs#L21-L27

So I think this solves the park/unpark reordering problem. Otherwise I don't think a compiler fence would completely fix it since the memory accesses could still have been reordered.

> I think Release might make sense here. It's also free on x86-64, and what they use in the corresponding part of the example: https://doc.rust-lang.org/std/thread/fn.park.html#examples

I've changed it to be `Release`.